include/mbedtls/ssl.h

```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
 #endif
 };
 
+/*
+ * Identifiers for PRFs used in various versions of TLS.
+ */
+typedef enum
+{
+   MBEDTLS_SSL_TLS_PRF_NONE,
+   MBEDTLS_SSL_TLS_PRF_SHA384,
+   MBEDTLS_SSL_TLS_PRF_SHA256,
+   MBEDTLS_SSL_HKDF_EXPAND_SHA384,
+   MBEDTLS_SSL_HKDF_EXPAND_SHA256
+}
+mbedtls_tls_prf_types;
+
+#if defined(MBEDTLS_SSL_EXPORT_KEYS)
+typedef enum
+{
+    MBEDTLS_SSL_KEY_EXPORT_TLS12_MASTER_SECRET = 0,
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL)
+    MBEDTLS_SSL_KEY_EXPORT_TLS13_CLIENT_EARLY_SECRET,
+    MBEDTLS_SSL_KEY_EXPORT_TLS13_EARLY_EXPORTER_SECRET,
+    MBEDTLS_SSL_KEY_EXPORT_TLS13_CLIENT_HANDSHAKE_TRAFFIC_SECRET,
+    MBEDTLS_SSL_KEY_EXPORT_TLS13_SERVER_HANDSHAKE_TRAFFIC_SECRET,
+    MBEDTLS_SSL_KEY_EXPORT_TLS13_CLIENT_APPLICATION_TRAFFIC_SECRET,
+    MBEDTLS_SSL_KEY_EXPORT_TLS13_SERVER_APPLICATION_TRAFFIC_SECRET,
+#endif /* MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL */
+} mbedtls_ssl_key_export_type;
+
+/**
+ * \brief           Callback type: Export key alongside random values for
+ *                                 session identification, and PRF for
+ *                                 implementation of TLS key exporters.
+ *
```

Comment:
I suggest to make the configuration SSL context specific instead, instead of passing the callback. This provides more flexibility to users.

Reply:
I don't like to add getters for the key material because then you have to explain if/when is it save to call them, e.g. what happens if you call them when inside/outside a handshake, or a renegotiation, ... It's much easier to directly pass the key material when it's ready.

Reply:
I don't think making the callback _function_ connection-specific is particularly useful. On the other hand, making the callback _context_ would be.

> you have to explain if/when is it save to call them

For the purpose here it's enough to guarantee that they're available during the call to the export callback.

Reply:
Why add a global getter which must only be called from the key export callback, instead of adding a parameter to the callback? That doesn't make sense to me.

Reply:
> I don't think making the callback function connection-specific is particularly useful. On the other hand, making the callback context would be.

I'd be OK with making the callback function itself configuration-specific, but the callback context connection-specific.

Reply:
> Why add a global getter which must only be called from the key export callback, instead of adding a parameter to the callback?

Because if (when) we later discover that it's useful to access some other data in the callback, we can add a getter function. We can't add a parameter to the callback.

Reply:
Note that the new design allows to add more elements to the enum. And if that happens not be enough (unlikely, IMO), we can still add getters which users can access using a reference to the SSL context which is part of the callback context.

Reply:
I have made the callback and context connection-specific for now. I agree that the callback itself could be configuration specific, but I prefer to keep things together here.

Reply:
I'm not convinced that it's better to pass all these arguments instead of letting the callback query them explicitly. But since the callback has access to the connection context, the API is extensible by adding getter functions, so the API is ok.

Reply:
After reviewing the discussions in #2188, #4363 and here, my gut feeling is it's unlikely but not impossible that the callback will need access to more information in the future, and I think the current API matches that nicely: it's optimized for the likely case when the callback needs what it currently does, by passing that as parameters (avoiding the need for getter functions), but also support extensions with getters should they be needed.